wordpress-mobile/WordPress-Android
# [Deps] Update Kotlin to 2.0.21

Project Thread: paaHJt-7n2-p2
Associated Metrics: paaHJt-7n2-p2#comment-9668

This PR updates Kotlin to 2.0 ([2.0.21](https://href.li/?https://github.com/JetBrains/kotlin/releases/tag/v2.0.21)).

## Description:

FYI: I recommend reviewing this PR commit-by-commit just to make sure that Kotlin has been updated appropriately and that there isn't any left-overs that need addressing.

PS: Main commits for some extra focus:
- [Build: Update/Add jvm target & source/target compatibility to java 11](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/6abe01ec983d0765d4bd1466030097109867ac56)
- [Analysis: Resolve val cannot be reassigned error on site store](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/7a9e63fef94728ac0046ff40ec2901f56e66c01b)
- [Analysis: Resolve non-public primary ctor warning on wordpress module](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/0ed041bf7b398c916fff41b22bb56156d9864627)
- [Deps: Update kotlin compile testing to 1.6.0 (tschuchortdev)](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/459eaa8a3d7cfc1214ba3641f45ba7bc4c82346e) (*)

(*) About `kotlin-compile-testing` ([tschuchortdev](https://github.com/tschuchortdev/kotlin-compile-testing)), it made me wonder how well maintained this library is and whether we can depend on in for future Kotlin updates, let's discuss. 🤔

-----

## To Test:



- Smoke test both, the `Jetpack` and `WordPress` apps, and verify everything is working as expected.
- Try to test for build time increases both, locally and on CI.
- You could also run the `./gradlew buildHealth` task locally and verify that everything is working as expected ([CI](https://buildkite.com/automattic/wordpress-android/builds/20336)).
- Review the [project dependencies & build environment changes](https://github.com/wordpress-mobile/WordPress-Android/pull/21428#issuecomment-2460220867) and assert correctness.

-----

## Regression Notes

1. Potential unintended areas of impact

    - Build time increases (clean and/or incremental).

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

3. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR Submission Checklist:


-----

## Testing Checklist (strike-out the not-applying and unnecessary ones): N/A
